File: erlang/allergies/src/allergies.erl
-module(allergies).
-export([test_version/0, allergies/1, is_allergic_to/2]).

-define(ALLERGENS, [{'eggs', 1}, 
                    {'peanuts', 2}, 
                    {'shellfish', 4}, 
                    {'strawberries', 8}, 
                    {'tomatoes', 16}, 
                    {'chocolate', 32}, 
                    {'pollen', 64}, 
                    {'cats', 128}]).

test_version() -> 1.

allergies(AllergyNumber) -> [Allergen || {Allergen, AllergenCode} <- ?ALLERGENS, AllergyNumber band AllergenCode == AllergenCode].

is_allergic_to(Allergen, AllergyNumber) -> lists:member(Allergen, allergies(AllergyNumber)).

File: csharp/allergies/Allergies.cs
using System.Linq;
using System.Collections.Generic;

public enum Allergen
{
    Eggs,  // 1
    Peanuts,  // 2
    Shellfish,  // 4
    Strawberries, // 8
    Tomatoes,  // 16
    Chocolate,  // 32
    Pollen,  // 64
    Cats  // 128
}

public class Allergies
{

    private Dictionary<Allergen, bool> allergies = new Dictionary<Allergen, bool>();

    public Allergies(int mask) {
        int maskCopy = mask;
        foreach (Allergen allergen in Allergen.GetValues(typeof(Allergen))) {
            CheckAllergy(maskCopy, allergen);
            maskCopy /= 2;
        }
    }

    private void CheckAllergy(int mask, Allergen allergen) {
        if (mask % 2 == 1) {
            allergies[allergen] = true;
        }
    }

    public bool IsAllergicTo(Allergen allergen)
    {
        return allergies.ContainsKey(allergen) && allergies[allergen];
    }

    public Allergen[] List()
    {
        return allergies.Keys.ToArray();
    }
}